protocol/xmrmaker/instance.go

```diff
@@ -57,15 +60,126 @@ func NewInstance(cfg *Config) (*Instance, error) {
 		go cfg.Network.Advertise()
 	}
 
-	return &Instance{
+	inst := &Instance{
 		backend:        cfg.Backend,
 		dataDir:        cfg.DataDir,
 		walletFile:     cfg.WalletFile,
 		walletPassword: cfg.WalletPassword,
 		offerManager:   om,
 		swapStates:     make(map[types.Hash]*swapState),
 		net:            cfg.Network,
-	}, nil
+	}
+
+	err = inst.checkForOngoingSwaps()
+	if err != nil {
+		return nil, err
+	}
+
+	return inst, nil
+}
+
+func (b *Instance) checkForOngoingSwaps() error {
+	swaps, err := b.backend.SwapManager().GetOngoingSwaps()
+	if err != nil {
+		return err
+	}
+
+	for _, s := range swaps {
+		if s.Provides != types.ProvidesXMR {
+			continue
+		}
+
+		if s.Status == types.KeysExchanged || s.Status == types.ExpectingKeys {
+			// TODO: set status to aborted, delete info from recovery db
+			continue
+		}
+
+		err = b.createOngoingSwap(s)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (b *Instance) createOngoingSwap(s *swap.Info) error {
+	// check if we have shared secret key in db; if so, recover XMR from that
+	// otherwise, create new swap state from recovery info
+	moneroStartHeight, err := b.backend.RecoveryDB().GetMoneroStartHeight(s.ID)
```

Comment:
I was assuming we were going to place the Monero start-scanning height in the `swap.Info` struct. Was there an advantage to storing it separately?

Reply:
I guess the same question applies to the relayer further down.

Reply:
honestly good question, I guess in my head I was thinking that recovery-specifc info would be stored separately, but I guess start heights/relayer info would make sense as part of `*swap.Info` also, since they're constant anyways.

Reply:
ah, I guess I didn't add the relayer info since it's specific to the xmrmaker. will leave that as-is but add the start heights to `*swap.Info`